spring-cloud-sleuth-instrumentation/src/main/java/org/springframework/cloud/sleuth/instrument/batch/TraceStepExecutionListener.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.batch;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import org.springframework.batch.core.ExitStatus;
+import org.springframework.batch.core.StepExecution;
+import org.springframework.batch.core.StepExecutionListener;
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.SpanAndScope;
+import org.springframework.cloud.sleuth.Tracer;
+
+class TraceStepExecutionListener implements StepExecutionListener {
+
+	private final Tracer tracer;
+
+	private static Map<StepExecution, SpanAndScope> SPANS = new ConcurrentHashMap<>();
+
+	TraceStepExecutionListener(Tracer tracer) {
+		this.tracer = tracer;
+	}
+
+	@Override
+	public void beforeStep(StepExecution stepExecution) {
+		Span span = this.tracer.nextSpan().name(stepExecution.getStepName());
+		span.tag("step.name", stepExecution.getStepName());
+		// TODO: How to add step type?
```

Comment:
Each `Step` implementation in Spring Batch adds a key named `batch.stepType` to the execution context with the implementation's class name as value (Here is an [example](https://github.com/spring-projects/spring-batch/blob/d8fc58338d3b059b67b5f777adc132d2564d7402/spring-batch-core/src/main/java/org/springframework/batch/core/partition/support/PartitionStep.java#L103)). However, this key is not available at this point as the listener is invoked before the actual execution of the step. A `StepExecution` object, which is the only object in scope here, does not provide a reference to the enclosing `Step`, but only its name. You can use a [StepLocator](https://docs.spring.io/spring-batch/docs/4.3.x/api/org/springframework/batch/core/step/StepLocator.html) to locate a step by name.

Reply:
Yes you should be able to get it from the execution context with the same [key](https://github.com/spring-projects/spring-batch/blob/d8fc58338d3b059b67b5f777adc132d2564d7402/spring-batch-core/src/main/java/org/springframework/batch/core/Step.java#L27) after the step is executed.